crypto/merklekeystore/keystore_test.go

```diff
@@ -0,0 +1,538 @@
+	}
+
+	_, err = signer.keyStore.GetKey(101)
+	a.Error(err)
+
+	signer = generateTestSigner(crypto.FalconType, 1000, 1100, 1, a)
+	defer signer.keyStore.store.Close()
+	for i := uint64(1000); i < 1100; i++ {
+		_, err = signer.keyStore.GetKey(i)
+		a.NoError(err, i)
+	}
+
+	_, err = signer.keyStore.GetKey(999)
+	a.Error(err)
+
+	signer = generateTestSigner(crypto.FalconType, 1000, 1100, 101, a)
+	intervalRounds := make([]uint64, 0)
+	for i := uint64(1000); i <= 1100; i++ {
+		if i%101 == 0 {
+			intervalRounds = append(intervalRounds, i)
+			continue
+		}
+		_, err := signer.keyStore.GetKey(i)
+		a.Error(err, i)
+	}
+}
+
+func TestNonEmptyDisposableKeys(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	a := require.New(t)
+
+	signer := generateTestSigner(crypto.FalconType, 0, 100, 1, a)
+	defer signer.keyStore.store.Close()
+
+	s := crypto.GenericSigningKey{}
+	for i := uint64(1); i <= 100; i++ {
+		key, err := signer.keyStore.GetKey(i)
+		a.NoError(err)
+		a.NotEqual(s, key)
+	}
+
+	signer = generateTestSigner(crypto.Ed25519Type, 0, 100, 1, a)
+	defer signer.keyStore.store.Close()
+
+	s = crypto.GenericSigningKey{}
+	for i := uint64(1); i <= 100; i++ {
+		key, err := signer.keyStore.GetKey(i)
+		a.NoError(err)
+		a.NotEqual(s, key)
+	}
+}
+
+func TestSignatureStructure(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	a := require.New(t)
+
+	signer := generateTestSigner(crypto.FalconType, 50, 100, 1, a)
+	defer signer.keyStore.store.Close()
+
+	hashable := genHashableForTest()
+	sig, err := signer.Sign(hashable, 51)
+	a.NoError(err)
+
+	key, err := signer.keyStore.GetKey(51)
+	a.NoError(err)
+
+	a.Equal(sig.VerifyingKey, *key.GetSigner().GetVerifyingKey())
+
+	proof, err := signer.Tree.Prove([]uint64{1})
+	a.NoError(err)
+	a.Equal(Proof(*proof), sig.Proof)
+
+	a.NotEqual(nil, sig.ByteSignature)
+}
+
+func genHashableForTest() crypto.Hashable {
+	hashable := TestingHashable{[]byte("test msg")}
+
+	return hashable
+}
+
+func TestSigning(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	a := require.New(t)
+
+	start, end := uint64(50), uint64(100)
+	signer := generateTestSigner(crypto.FalconType, start, end, 1, a)
+	defer signer.keyStore.store.Close()
+
+	hashable := genHashableForTest()
+
+	sig, err := signer.Sign(hashable, start)
+	a.NoError(err)
+	a.NoError(signer.GetVerifier().Verify(start, start, 1, hashable, sig))
+
+	_, err = signer.Sign(hashable, start-1)
+	a.Error(err)
+
+	_, err = signer.Sign(hashable, end+1)
+	a.Error(err)
+
+	signer = generateTestSigner(crypto.FalconType, start, end, 10, a)
+	defer signer.keyStore.store.Close()
+
+	sig, err = signer.Sign(hashable, start)
+	a.NoError(err)
+	a.NoError(signer.GetVerifier().Verify(start, start, 1, hashable, sig))
+
+	sig, err = signer.Sign(hashable, start+5)
+	a.Error(err)
+	a.Error(signer.GetVerifier().Verify(start, start+5, 1, hashable, sig))
+
+	signer = generateTestSigner(crypto.FalconType, 50, 100, 12, a)
+	defer signer.keyStore.store.Close()
+	a.Equal(length(signer, a), 4)
```

Comment:
`Equal` first argument is the expected value, second is the checked value. The parameter order must be reversed.